libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -35,3 +38,95 @@ string IRGenerationContext::addLocalVariable(VariableDeclaration const& _varDecl
+		"Unknown variable: " + _varDecl.name()
+	);
+	return m_localVariables[&_varDecl];
+}
+
+string IRGenerationContext::functionName(FunctionDefinition const& _function)
+{
+	// @TODO previously, we had to distinguish creation context and runtime context,
+	// but since we do not work with jump positions anymore, this should not be a problem, right?
+	return "fun_" + _function.name() + "_" + to_string(_function.id());
+}
+
+FunctionDefinition const& IRGenerationContext::virtualFunction(FunctionDefinition const& _function)
+{
+	// @TODO previously, we had to distinguish creation context and runtime context,
+	// but since we do not work with jump positions anymore, this should not be a problem, right?
+	string name = _function.name();
+	FunctionType functionType(_function);
+	for (auto const& contract: m_inheritanceHierarchy)
+		for (FunctionDefinition const* function: contract->definedFunctions())
+			if (
+				function->name() == name &&
+				!function->isConstructor() &&
+				FunctionType(*function).asCallableFunction(false)->hasEqualParameterTypes(functionType)
+			)
+				return *function;
+	solAssert(false, "Super function " + name + " not found.");
+}
+
+string IRGenerationContext::virtualFunctionName(FunctionDefinition const& _functionDeclaration)
+{
+	return functionName(virtualFunction(_functionDeclaration));
+}
+
+string IRGenerationContext::newYulVariable()
+{
+	return "_" + to_string(++m_varCounter);
+}
+
+string IRGenerationContext::variable(Expression const& _expression)
+{
+	unsigned size = _expression.annotation().type->sizeOnStack();
+	solUnimplementedAssert(size == 1, "");
+	return "expr_" + to_string(_expression.id());
+}
+
+string IRGenerationContext::internalDispatch(size_t _in, size_t _out)
+{
+	// TODO can we limit the generated functions to only those visited
+	// in the expression context? What about creation / runtime context?
+	string funName = "dispatch_internal_in_" + to_string(_in) + "_out_" + to_string(_out);
+	return m_functions->createFunction(funName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(fun <comma> <in>) -> <out> {
+				switch fun
+				<#cases>
+				case <funID>
+				{
+					<out> := <name>(<in>)
+				}
+				</cases>
+				default { invalid() }
+			}
+		)");
+		templ("functionName", funName);
+		templ("comma", _in > 0 ? "," : "");
+		YulUtilFunctions utils(m_evmVersion, m_functions);
+		templ("in", utils.suffixedVariableNameList("in_", 0, _in));
+		templ("out", utils.suffixedVariableNameList("out_", 0, _out));
+		vector<map<string, string>> functions;
+		for (auto const& contract: m_inheritanceHierarchy)
+			for (FunctionDefinition const* function: contract->definedFunctions())
+				if (
+					!function->isConstructor() &&
+					function->parameters().size() == _in &&
+					function->returnParameters().size() == _out
+				)
+				{
+					map<string, string> f;
+					f["funID"] = to_string(function->id());
+					f["name"] = functionName(*function);
+					functions.emplace_back(move(f));
```

Comment:
```suggestion
					functions.emplace_back(map<string, string> {
						{ "funID", to_string(function->id()) },
						{ "name", functionName(*function) }
					});
```
? Could even loose the braces then.

Reply:
Sounds good!